packages/astro/src/content/vite-plugin-content-virtual-mod.ts

```diff
@@ -20,98 +21,177 @@ import {
+		const contentEntryConfigByExt = getEntryConfigByExtMap(settings.contentEntryTypes);
+		const contentEntryExts = [...contentEntryConfigByExt.keys()];
+		const dataEntryExts = getDataEntryExts(settings);
+		const createGlob = (value: string[], flag: string) => `import.meta.glob(${JSON.stringify(value)}, { query: { ${flag}: true } })`
+		contentEntryGlobResult = createGlob(globWithUnderscoresIgnored(relContentDir, contentEntryExts), CONTENT_FLAG);
+		dataEntryGlobResult = createGlob(globWithUnderscoresIgnored(relContentDir, dataEntryExts), DATA_FLAG);
+		renderEntryGlobResult = createGlob(globWithUnderscoresIgnored(relContentDir, contentEntryExts), CONTENT_RENDER_FLAG);
+	} else {
+		contentEntryGlobResult = getStringifiedCollectionFromLookup('content', relContentDir, lookupMap);
+		dataEntryGlobResult = getStringifiedCollectionFromLookup('data', relContentDir, lookupMap);
+		renderEntryGlobResult = getStringifiedCollectionFromLookup('render', relContentDir, lookupMap);
+	}
 
-	const virtualModContents = fsMod
+	const virtualModContents = nodeFs
 		.readFileSync(contentPaths.virtualModTemplate, 'utf-8')
-		.replace(
-			'@@COLLECTION_NAME_BY_REFERENCE_KEY@@',
-			new URL('reference-map.json', contentPaths.cacheDir).pathname
-		)
 		.replace('@@CONTENT_DIR@@', relContentDir)
 		.replace(
 			"'@@CONTENT_ENTRY_GLOB_PATH@@'",
-			JSON.stringify(globWithUnderscoresIgnored(relContentDir, contentEntryExts))
+			contentEntryGlobResult
 		)
 		.replace(
 			"'@@DATA_ENTRY_GLOB_PATH@@'",
-			JSON.stringify(globWithUnderscoresIgnored(relContentDir, dataEntryExts))
+			dataEntryGlobResult
 		)
 		.replace(
 			"'@@RENDER_ENTRY_GLOB_PATH@@'",
-			JSON.stringify(
-				globWithUnderscoresIgnored(
-					relContentDir,
-					/** Note: data collections excluded */ contentEntryExts
-				)
-			)
+			renderEntryGlobResult
+		).replace(
+			'/* @@LOOKUP_MAP_ASSIGNMENT@@ */',
+			`lookupMap = ${JSON.stringify(lookupMap)};`
 		);
 
-	const astroContentVirtualModuleId = '\0' + VIRTUAL_MODULE_ID;
-
-	return {
-		name: 'astro-content-virtual-mod-plugin',
-		resolveId(id) {
-			if (id === VIRTUAL_MODULE_ID) {
-				return astroContentVirtualModuleId;
-			}
-		},
-		async load(id) {
-			if (id === astroContentVirtualModuleId) {
-				const stringifiedLookupMap = await getStringifiedLookupMap({
-					fs: fsMod,
-					contentPaths,
-					contentEntryConfigByExt,
-					dataEntryExts,
-					root: settings.config.root,
-				});
+	return virtualModContents;
+}
 
-				return {
-					code: virtualModContents.replace(
-						'/* @@LOOKUP_MAP_ASSIGNMENT@@ */',
-						`lookupMap = ${stringifiedLookupMap};`
-					),
-				};
+function getStringifiedCollectionFromLookup(wantedType: 'content' | 'data' | 'render', relContentDir: string, lookupMap: ContentLookupMap) {
+	let str = '{';
+	// In dev, we don't need to normalize the import specifier at all. Vite handles it.
+	let normalize = (slug: string) => slug;
```

Comment:
This can be `let normalize` because `if/else` both reach the point where we do `normalize =`